File: src/bigclaw/issue_archive.py
from dataclasses import dataclass, field
from typing import Dict, List


ALLOWED_ISSUE_CATEGORIES = {"ui", "ia", "permission", "metric"}
ALLOWED_ISSUE_PRIORITIES = {"P0", "P1", "P2"}


@dataclass(frozen=True)
class ArchivedIssue:
    finding_id: str
    summary: str
    category: str
    priority: str
    owner: str
    surface: str = ""
    impact: str = ""
    status: str = "open"
    evidence: List[str] = field(default_factory=list)

    @property
    def normalized_category(self) -> str:
        return self.category.strip().lower()

    @property
    def normalized_priority(self) -> str:
        return self.priority.strip().upper()

    @property
    def resolved(self) -> bool:
        return self.status.strip().lower() == "resolved"

    def to_dict(self) -> Dict[str, object]:
        return {
            "finding_id": self.finding_id,
            "summary": self.summary,
            "category": self.category,
            "priority": self.priority,
            "owner": self.owner,
            "surface": self.surface,
            "impact": self.impact,
            "status": self.status,
            "evidence": list(self.evidence),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "ArchivedIssue":
        return cls(
            finding_id=str(data["finding_id"]),
            summary=str(data["summary"]),
            category=str(data["category"]),
            priority=str(data["priority"]),
            owner=str(data.get("owner", "")),
            surface=str(data.get("surface", "")),
            impact=str(data.get("impact", "")),
            status=str(data.get("status", "open")),
            evidence=[str(item) for item in data.get("evidence", [])],
        )


@dataclass
class IssuePriorityArchive:
    issue_id: str
    title: str
    version: str
    findings: List[ArchivedIssue] = field(default_factory=list)

    def to_dict(self) -> Dict[str, object]:
        return {
            "issue_id": self.issue_id,
            "title": self.title,
            "version": self.version,
            "findings": [finding.to_dict() for finding in self.findings],
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "IssuePriorityArchive":
        return cls(
            issue_id=str(data["issue_id"]),
            title=str(data["title"]),
            version=str(data["version"]),
            findings=[ArchivedIssue.from_dict(item) for item in data.get("findings", [])],
        )


@dataclass(frozen=True)
class IssuePriorityArchiveAudit:
    ready: bool
    finding_count: int
    priority_counts: Dict[str, int] = field(default_factory=dict)
    category_counts: Dict[str, int] = field(default_factory=dict)
    missing_owners: List[str] = field(default_factory=list)
    invalid_priorities: List[str] = field(default_factory=list)
    invalid_categories: List[str] = field(default_factory=list)
    unresolved_p0_findings: List[str] = field(default_factory=list)

    @property
    def summary(self) -> str:
        status = "READY" if self.ready else "HOLD"
        return (
            f"{status}: findings={self.finding_count} "
            f"missing_owners={len(self.missing_owners)} "
            f"invalid_priorities={len(self.invalid_priorities)} "
            f"invalid_categories={len(self.invalid_categories)} "
            f"unresolved_p0={len(self.unresolved_p0_findings)}"
        )

    def to_dict(self) -> Dict[str, object]:
        return {
            "ready": self.ready,
            "finding_count": self.finding_count,
            "priority_counts": dict(self.priority_counts),
            "category_counts": dict(self.category_counts),
            "missing_owners": list(self.missing_owners),
            "invalid_priorities": list(self.invalid_priorities),
            "invalid_categories": list(self.invalid_categories),
            "unresolved_p0_findings": list(self.unresolved_p0_findings),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "IssuePriorityArchiveAudit":
        return cls(
            ready=bool(data["ready"]),
            finding_count=int(data.get("finding_count", 0)),
            priority_counts={
                str(priority): int(count)
                for priority, count in dict(data.get("priority_counts", {})).items()
            },
            category_counts={
                str(category): int(count)
                for category, count in dict(data.get("category_counts", {})).items()
            },
            missing_owners=[str(item) for item in data.get("missing_owners", [])],
            invalid_priorities=[str(item) for item in data.get("invalid_priorities", [])],
            invalid_categories=[str(item) for item in data.get("invalid_categories", [])],
            unresolved_p0_findings=[str(item) for item in data.get("unresolved_p0_findings", [])],
        )


class IssuePriorityArchivist:
    def audit(self, archive: IssuePriorityArchive) -> IssuePriorityArchiveAudit:
        priority_counts = {priority: 0 for priority in sorted(ALLOWED_ISSUE_PRIORITIES)}
        category_counts = {category: 0 for category in sorted(ALLOWED_ISSUE_CATEGORIES)}
        missing_owners: List[str] = []
        invalid_priorities: List[str] = []
        invalid_categories: List[str] = []
        unresolved_p0_findings: List[str] = []

        for finding in archive.findings:
            if not finding.owner.strip():
                missing_owners.append(finding.finding_id)

            if finding.normalized_priority in ALLOWED_ISSUE_PRIORITIES:
                priority_counts[finding.normalized_priority] += 1
            else:
                invalid_priorities.append(finding.finding_id)

            if finding.normalized_category in ALLOWED_ISSUE_CATEGORIES:
                category_counts[finding.normalized_category] += 1
            else:
                invalid_categories.append(finding.finding_id)

            if finding.normalized_priority == "P0" and not finding.resolved:
                unresolved_p0_findings.append(finding.finding_id)

        ready = bool(archive.findings) and not (
            missing_owners or invalid_priorities or invalid_categories or unresolved_p0_findings
        )
        return IssuePriorityArchiveAudit(
            ready=ready,
            finding_count=len(archive.findings),
            priority_counts=priority_counts,
            category_counts=category_counts,
            missing_owners=sorted(missing_owners),
            invalid_priorities=sorted(invalid_priorities),
            invalid_categories=sorted(invalid_categories),
            unresolved_p0_findings=sorted(unresolved_p0_findings),
        )


def render_issue_priority_archive_report(
    archive: IssuePriorityArchive,
    audit: IssuePriorityArchiveAudit,
) -> str:
    lines = [
        "# Issue Priority Archive",
        "",
        f"- Issue: {archive.issue_id} {archive.title}",
        f"- Version: {archive.version}",
        f"- Audit: {audit.summary}",
        (
            "- Priority Counts: "
            f"P0={audit.priority_counts.get('P0', 0)} "
            f"P1={audit.priority_counts.get('P1', 0)} "
            f"P2={audit.priority_counts.get('P2', 0)}"
        ),
        (
            "- Category Counts: "
            f"ui={audit.category_counts.get('ui', 0)} "
            f"ia={audit.category_counts.get('ia', 0)} "
            f"permission={audit.category_counts.get('permission', 0)} "
            f"metric={audit.category_counts.get('metric', 0)}"
        ),
        "",
        "## Findings",
    ]
    for finding in archive.findings:
        lines.append(
            "- "
            f"{finding.finding_id}: {finding.summary} "
            f"category={finding.normalized_category} priority={finding.normalized_priority} "
            f"owner={finding.owner or 'none'} status={finding.status}"
        )
        lines.append(
            "  "
            f"surface={finding.surface or 'none'} impact={finding.impact or 'none'} "
            f"evidence={','.join(finding.evidence) or 'none'}"
        )

    lines.extend(
        [
            "",
            "## Audit Findings",
            f"- Missing owners: {', '.join(audit.missing_owners) or 'none'}",
            f"- Invalid priorities: {', '.join(audit.invalid_priorities) or 'none'}",
            f"- Invalid categories: {', '.join(audit.invalid_categories) or 'none'}",
            f"- Unresolved P0 findings: {', '.join(audit.unresolved_p0_findings) or 'none'}",
        ]
    )
    return "\n".join(lines)

File: bigclaw-go/internal/product/issue_archive.go
package product

import (
	"encoding/json"
	"fmt"
	"slices"
	"strings"
)

var allowedIssueCategories = []string{"ia", "metric", "permission", "ui"}
var allowedIssuePriorities = []string{"P0", "P1", "P2"}

type ArchivedIssue struct {
	FindingID string   `json:"finding_id"`
	Summary   string   `json:"summary"`
	Category  string   `json:"category"`
	Priority  string   `json:"priority"`
	Owner     string   `json:"owner,omitempty"`
	Surface   string   `json:"surface,omitempty"`
	Impact    string   `json:"impact,omitempty"`
	Status    string   `json:"status,omitempty"`
	Evidence  []string `json:"evidence,omitempty"`
}

func (issue ArchivedIssue) NormalizedCategory() string {
	return strings.ToLower(strings.TrimSpace(issue.Category))
}

func (issue ArchivedIssue) NormalizedPriority() string {
	return strings.ToUpper(strings.TrimSpace(issue.Priority))
}

func (issue ArchivedIssue) Resolved() bool {
	return strings.EqualFold(strings.TrimSpace(issue.Status), "resolved")
}

func (issue ArchivedIssue) MarshalJSON() ([]byte, error) {
	payload := map[string]any{
		"finding_id": issue.FindingID,
		"summary":    issue.Summary,
		"category":   issue.Category,
		"priority":   issue.Priority,
		"owner":      issue.Owner,
		"surface":    issue.Surface,
		"impact":     issue.Impact,
		"status":     archivedIssueStatus(issue.Status),
		"evidence":   issueArchiveStrings(issue.Evidence),
	}
	return json.Marshal(payload)
}

func (issue *ArchivedIssue) UnmarshalJSON(data []byte) error {
	type alias ArchivedIssue
	aux := struct {
		Status *string `json:"status"`
		*alias
	}{
		alias: (*alias)(issue),
	}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if aux.Status == nil || strings.TrimSpace(*aux.Status) == "" {
		issue.Status = "open"
	} else {
		issue.Status = *aux.Status
	}
	if issue.Evidence == nil {
		issue.Evidence = []string{}
	}
	return nil
}

type IssuePriorityArchive struct {
	IssueID  string          `json:"issue_id"`
	Title    string          `json:"title"`
	Version  string          `json:"version"`
	Findings []ArchivedIssue `json:"findings,omitempty"`
}

func (archive IssuePriorityArchive) MarshalJSON() ([]byte, error) {
	payload := map[string]any{
		"issue_id": archive.IssueID,
		"title":    archive.Title,
		"version":  archive.Version,
		"findings": issueArchiveFindings(archive.Findings),
	}
	return json.Marshal(payload)
}

func (archive *IssuePriorityArchive) UnmarshalJSON(data []byte) error {
	type alias IssuePriorityArchive
	aux := struct{ *alias }{alias: (*alias)(archive)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if archive.Findings == nil {
		archive.Findings = []ArchivedIssue{}
	}
	return nil
}

type IssuePriorityArchiveAudit struct {
	Ready                bool           `json:"ready"`
	FindingCount         int            `json:"finding_count"`
	PriorityCounts       map[string]int `json:"priority_counts,omitempty"`
	CategoryCounts       map[string]int `json:"category_counts,omitempty"`
	MissingOwners        []string       `json:"missing_owners,omitempty"`
	InvalidPriorities    []string       `json:"invalid_priorities,omitempty"`
	InvalidCategories    []string       `json:"invalid_categories,omitempty"`
	UnresolvedP0Findings []string       `json:"unresolved_p0_findings,omitempty"`
}

func (audit IssuePriorityArchiveAudit) Summary() string {
	status := "HOLD"
	if audit.Ready {
		status = "READY"
	}
	return fmt.Sprintf(
		"%s: findings=%d missing_owners=%d invalid_priorities=%d invalid_categories=%d unresolved_p0=%d",
		status,
		audit.FindingCount,
		len(audit.MissingOwners),
		len(audit.InvalidPriorities),
		len(audit.InvalidCategories),
		len(audit.UnresolvedP0Findings),
	)
}

func (audit IssuePriorityArchiveAudit) MarshalJSON() ([]byte, error) {
	payload := map[string]any{
		"ready":                  audit.Ready,
		"finding_count":          audit.FindingCount,
		"priority_counts":        issueArchiveCounts(audit.PriorityCounts),
		"category_counts":        issueArchiveCounts(audit.CategoryCounts),
		"missing_owners":         issueArchiveStrings(audit.MissingOwners),
		"invalid_priorities":     issueArchiveStrings(audit.InvalidPriorities),
		"invalid_categories":     issueArchiveStrings(audit.InvalidCategories),
		"unresolved_p0_findings": issueArchiveStrings(audit.UnresolvedP0Findings),
	}
	return json.Marshal(payload)
}

func (audit *IssuePriorityArchiveAudit) UnmarshalJSON(data []byte) error {
	type alias IssuePriorityArchiveAudit
	aux := struct{ *alias }{alias: (*alias)(audit)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if audit.PriorityCounts == nil {
		audit.PriorityCounts = map[string]int{}
	}
	if audit.CategoryCounts == nil {
		audit.CategoryCounts = map[string]int{}
	}
	if audit.MissingOwners == nil {
		audit.MissingOwners = []string{}
	}
	if audit.InvalidPriorities == nil {
		audit.InvalidPriorities = []string{}
	}
	if audit.InvalidCategories == nil {
		audit.InvalidCategories = []string{}
	}
	if audit.UnresolvedP0Findings == nil {
		audit.UnresolvedP0Findings = []string{}
	}
	return nil
}

type IssuePriorityArchivist struct{}

func (IssuePriorityArchivist) Audit(archive IssuePriorityArchive) IssuePriorityArchiveAudit {
	priorityCounts := map[string]int{}
	for _, priority := range allowedIssuePriorities {
		priorityCounts[priority] = 0
	}
	categoryCounts := map[string]int{}
	for _, category := range allowedIssueCategories {
		categoryCounts[category] = 0
	}

	var missingOwners []string
	var invalidPriorities []string
	var invalidCategories []string
	var unresolvedP0 []string

	for _, finding := range archive.Findings {
		if strings.TrimSpace(finding.Owner) == "" {
			missingOwners = append(missingOwners, finding.FindingID)
		}
		if priority := finding.NormalizedPriority(); slices.Contains(allowedIssuePriorities, priority) {
			priorityCounts[priority]++
		} else {
			invalidPriorities = append(invalidPriorities, finding.FindingID)
		}
		if category := finding.NormalizedCategory(); slices.Contains(allowedIssueCategories, category) {
			categoryCounts[category]++
		} else {
			invalidCategories = append(invalidCategories, finding.FindingID)
		}
		if finding.NormalizedPriority() == "P0" && !finding.Resolved() {
			unresolvedP0 = append(unresolvedP0, finding.FindingID)
		}
	}

	slices.Sort(missingOwners)
	slices.Sort(invalidPriorities)
	slices.Sort(invalidCategories)
	slices.Sort(unresolvedP0)

	ready := len(archive.Findings) > 0 &&
		len(missingOwners) == 0 &&
		len(invalidPriorities) == 0 &&
		len(invalidCategories) == 0 &&
		len(unresolvedP0) == 0

	return IssuePriorityArchiveAudit{
		Ready:                ready,
		FindingCount:         len(archive.Findings),
		PriorityCounts:       priorityCounts,
		CategoryCounts:       categoryCounts,
		MissingOwners:        missingOwners,
		InvalidPriorities:    invalidPriorities,
		InvalidCategories:    invalidCategories,
		UnresolvedP0Findings: unresolvedP0,
	}
}

func RenderIssuePriorityArchiveReport(archive IssuePriorityArchive, audit IssuePriorityArchiveAudit) string {
	lines := []string{
		"# Issue Priority Archive",
		"",
		fmt.Sprintf("- Issue: %s %s", archive.IssueID, archive.Title),
		fmt.Sprintf("- Version: %s", archive.Version),
		fmt.Sprintf("- Audit: %s", audit.Summary()),
		fmt.Sprintf("- Priority Counts: P0=%d P1=%d P2=%d", audit.PriorityCounts["P0"], audit.PriorityCounts["P1"], audit.PriorityCounts["P2"]),
		fmt.Sprintf("- Category Counts: ui=%d ia=%d permission=%d metric=%d", audit.CategoryCounts["ui"], audit.CategoryCounts["ia"], audit.CategoryCounts["permission"], audit.CategoryCounts["metric"]),
		"",
		"## Findings",
	}
	for _, finding := range archive.Findings {
		lines = append(lines, fmt.Sprintf(
			"- %s: %s category=%s priority=%s owner=%s status=%s",
			finding.FindingID,
			finding.Summary,
			finding.NormalizedCategory(),
			finding.NormalizedPriority(),
			firstNonEmpty(strings.TrimSpace(finding.Owner), "none"),
			archivedIssueStatus(finding.Status),
		))
		lines = append(lines, fmt.Sprintf(
			"  surface=%s impact=%s evidence=%s",
			firstNonEmpty(strings.TrimSpace(finding.Surface), "none"),
			firstNonEmpty(strings.TrimSpace(finding.Impact), "none"),
			firstNonEmpty(strings.Join(finding.Evidence, ","), "none"),
		))
	}
	lines = append(lines,
		"",
		"## Audit Findings",
		fmt.Sprintf("- Missing owners: %s", firstNonEmpty(strings.Join(audit.MissingOwners, ", "), "none")),
		fmt.Sprintf("- Invalid priorities: %s", firstNonEmpty(strings.Join(audit.InvalidPriorities, ", "), "none")),
		fmt.Sprintf("- Invalid categories: %s", firstNonEmpty(strings.Join(audit.InvalidCategories, ", "), "none")),
		fmt.Sprintf("- Unresolved P0 findings: %s", firstNonEmpty(strings.Join(audit.UnresolvedP0Findings, ", "), "none")),
	)
	return strings.Join(lines, "\n")
}

func archivedIssueStatus(status string) string {
	trimmed := strings.TrimSpace(status)
	if trimmed == "" {
		return "open"
	}
	return trimmed
}

func issueArchiveStrings(values []string) []string {
	if values == nil {
		return []string{}
	}
	return values
}

func issueArchiveFindings(values []ArchivedIssue) []ArchivedIssue {
	if values == nil {
		return []ArchivedIssue{}
	}
	return values
}

func issueArchiveCounts(values map[string]int) map[string]int {
	if values == nil {
		return map[string]int{}
	}
	return values
}

func firstNonEmpty(value string, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	return value
}
